docs/proposals/generalize-modules.md

```diff
@@ -0,0 +1,72 @@
+---
+title: "Generalize Modules Service to make it extensible"
+linkTitle: "Generalize Modules Service to make it extensible"
+weight: 1
+slug: generalize-modules
+---
+
+- Author: @annanay25
+- Reviewers: 
+- Date: April 2020
+- Status: Draft
+
+## Overview
+
+Cortex uses modules to start and operate services with dependencies. Inter-service dependencies are specified in a map and passed to a module manager which ensures that they are started in the right order of dependencies. While this works really well, the implementation is tied in specifically to the Cortex struct and is not flexible for use with other projects like Loki, which also require similar forms of dependency management. 
+ 
+We would like to extend modules in cortex to a generic dependency management framework, that can be used by any project with no ties to cortex. 
+
+## Specific goals
+
+- Framework should allow for reusing cortex modules and allow us to:
+  - Add new modules
+  - Overwrite the implementation of a current module
+  - Manage dependencies
+- Framework should allow for building an application from scratch using the `modules` package, with no dependencies on Cortex. For ex: Remove code from Loki that was copied from `pkg/cortex/cortex.go`.
+
+
+
+## Proposed Design
+
+To make the modules package extensible, we need to abstract away any Cortex specific details from the module manager. The proposed design is to -
+
+- Make a new component `moduleManager`, which is envisioned to be a central manager for all modules of the application. It stores modules & dependencies, and will be housed under a new package `pkg/util/modules`. The following is the interface for interacting with the `moduleManager`:
+```
+// Manager defines the interface for interaction with moduleManager.
+type Manager interface {
+   RegisterModule(name string, deps []string, svc service, options ...Option)
+   AddDependency(fromModule string, toModule string) error
```

Comment:
I think we could remove this from the interface as well by adding a `SetDependant` option. When provided it can update the target module with the one currently being registered as a dependency.

Reply:
It could be useful to define dependencies on modules that are not yet registered. Possible errors should be detected later.